snowplow-incubator/snowplow-roku-tracker
# Initial release (close #1)

The tracker is implemented using [BrighterScript](https://github.com/rokucommunity/brighterscript) which is a superset of the official BrightScript scripting language for Roku. BrighterScript transpiles into BrightScript. In this way, it is similar to TypeScript and JavaScript.

The project is intended to be developed in Visual Studio Code which has a BrightScript extension. There are debug targets for running a demo app and tests in VS Code. These run npm scripts on the background to build the package and then upload it to the Roku device.

Before deploying the tracker, you should enable developer mode on your Roku as described in the official docs. Afterwards, you need to add your Roku device username and password to the `.vscode/.env` file and add configuration for your Snowplow Mini instance to manifest files for demo and tests. This is explained below in the demo and testing section.

Once ready, we should publish the package on NPM so that it can also be installed using the Roku Package Manager.

# Public Interface

The API of the tracker is a single component called "Snowplow". It is designed to resemble the official [Roku Analytics Component Library](https://developer.roku.com/en-gb/docs/developer-program/libraries/roku-analytics-component.md). The interface provides an `init` property to initialize a tracker given a configuration. In addition, it has properties for tracking the following types of events: screen view, structured and self-describing.

To use in apps, you should instantiate Snowplow and add it to the global scope.
In this way, it will be accessible from anywhere within your SceneGraph application.
You may create the instance in the `init` function of your main scene.

```
m.global.AddField("snowplow", "node", false)
m.global.snowplow = CreateObject("roSGNode", "Snowplow")
```

Trackers are initialized by setting the `init` property with configuration of the tracker.
Below is an example configuration, please refer to the documentation for a complete specification.

```
m.global.snowplow.init = {
    namespace: "ns1", ' optional, defaults to "default"
    network: {
        collector: "https://...",
        method: "POST",
        retryCount: 3 ' optional, defaults to 0
    },
    subject: {
        domainUserId: "",
        networkUserId: "",
        appId: ""
    }
}
```

You may initialize multiple trackers with different namespaces.
Setting the `init` property with the same tracker namespace updates the configuration of the tracker.
When no namespace is given, a namespace with the name "default" is used (both in configuration and when tracking events).

To track events, simply assign their properties as associative arrays to fields corresponding to event types.
Events will be sent to all initialized trackers (with any namespace) when setting the event types on the Snowplow instance.
To send events to a specific tracker, call its namespace as follows: `m.global.snowplow.trackerNamespace.structured = {...}`.
The following are some examples of the tracked events.
Please refer to the documentation for specificiation of event properties.

```
' Tracking a screen view event using all initialized trackers
m.global.snowplow.screenView = {
    id: CreateObject("roDeviceInfo").GetRandomUUID(),
    name: "HomeScene",
    type: "scene",
    transitionType: "init"
}

' Tracking a screen view event using a tracker with the namespace "ns1"
m.global.snowplow.ns1.screenView = {
    id: CreateObject("roDeviceInfo").GetRandomUUID(),
    name: "HomeScene",
    type: "scene",
    transitionType: "init"
}

' Tracking a self-describing events
m.global.snowplow.selfDescribing = {
    data: {
        id: CreateObject("roDeviceInfo").GetRandomUUID(),
        name: "name",
        type: "type"
    },
    schema: "iglu:com.snowplowanalytics.mobile/screen_view/jsonschema/1-0-0"
}

' Tracking a structured event
m.global.snowplow.structured = {
    namespace: "otherNamespace",
    se_ca: "category",
    se_ac: "action",
    se_la: "label",
    se_pr: "property",
    se_va: 10
}

' Adding context to an event
m.global.snowplow.structured = {
    ...
    contexts: [
        {
            data: {
                id: CreateObject("roDeviceInfo").GetRandomUUID(),
                name: "contextName",
                type: "type"
            },
            schema: "iglu:com.snowplowanalytics.mobile/screen_view/jsonschema/1-0-0"
        }
    ]
}
```

# Internal Architecture

The internal code of the tracker is fully implemented in BrighterScript and makes use of its OO-features. It is placed within a Snowplow.Internal namespace. It contains the following classes:

* Tracker
* Subject
* Emitter
* Configuration
* NetworkConnection
* Event, ScreenView, Structured, SelfDescribing, EventContext

[A UML class diagram can be accessed here](https://docs.google.com/drawings/d/1T9S4CpnkZPwBYcRPXdpNCGrJKIW8A170ogDgNRnhtKs/edit?usp=sharing).

Everything is executed in a background (non-render) thread. The thread listens to changes in the properties of the public interface and processes the requests.

# Demo App

The tracker comes with a demo app that shows it in use.
It is a simple video player with a fixed collection of played videos and an interface to switch between them.

The project is located in the `src-demo-app` subfolder.
The following steps assume that you have enabled developer mode on your Roku device and it is connected to your network.

1. Run `npm install` on the project
2. Install ropm `npm i ropm -g`
3. Install ropm packages `ropm install`
4. Create `.env` file with environment variables
```
ROKU_IP=192.168.100.129
ROKU_PASSWORD=XXXX
```
5. Add configuration for Snowplow collector to `src-demo-app/manifest`
```
snowplow_collector=http://192.168.100.127:9090
snowplow_method=POST
```
6. Start the demo app using `npm run demo-app`

Alternatively, you may run the demo app from Visual Studio Code as the debug configuration is already prepared.
Install the BrightScript extension to Visual Studio Code and choose "Run demo app" in the debug options.

Events will be sent to the Snowplow collector as you navigate through the app.

# Testing

The tracker functionality is verified using unit and integration tests.
Unit tests test individual components of the tracker in isolation and do not make any external network requests.
Integration tests use a Snowplow Micro instance to verify end-to-end tracking of events.

The tests are located in the `tests` subfolder.
The following steps assume that you have enabled developer mode on your Roku device and it is connected to your network.

1. Run `npm install` on the project
2. Install ropm `npm i ropm -g`
3. Install ropm packages `ropm install`
4. Create `.env` file with environment variables
```
ROKU_IP=192.168.100.129
ROKU_PASSWORD=XXXX
```
5. Start [Snowplow Micro](https://github.com/snowplow-incubator/snowplow-micro) on your computer
6. Add configuration for your Snowplow Micro (with the network IP of your computer) instance to `tests/manifest`
```
snowplow_collector=http://192.168.100.127:9090
snowplow_method=POST
```
7. Run the tests using `npm test`

Alternatively, you may run the tests from Visual Studio Code as the debug configuration is already prepared.
Install the BrightScript extension to Visual Studio Code and choose "Run tests" in the debug options.

# Limitations

The main limitations of this MVP of the tracker are:

* no local event store, events are sent right after they are received (there is a simple retry mechanism provided the roku-requests package)
* no additional context added to events – Roku enables to access a lot of information about the platform which is not used yet as it would require changes in our schemas
* no support for sessions
* no global contexts
* no documentation yet
